spec.bs

```diff
@@ -6,57 +6,940 @@ Status: CG-DRAFT
+            1. If |pendingBits| is greater than 0, run the following steps:
+                1. Let |ledger| be [=user agent=]'s [=shared storage budget table=][|origin|].
+                1. Let |bitDebit| be a new [=bit debit=].
+                1. Set |bitDebit|'s [=bit debit/bits=] to |pendingBits|.
+                1. Let |unsafeMoment| be the [=wall clock=]'s [=wall clock/unsafe current time=].
+                1. Let |moment| be the result of running [=coarsen time=] algorithm given |unsafeMoment| and [=wall clock=] as input.
+                1. Let |fromUnixEpochTime| be the [=duration from=] the [=Unix epoch=] to |moment|.
+                1. Set |bitDebit|'s [=bit debit/timestamp=] to |fromUnixEpochTime|.
+                1. [=list/Append=] |bitDebit| to |ledger|.
+                1. Set |node|'s {{FencedFrameConfig}}'s [=pending shared storage budget debit=] to 0.
+        1. Set |node| to |node|'s [=navigable/parent=].
+  </div>
+
+  ### Run Operation Methods ### {#run-op}
+
+  <div algorithm>
+    The <dfn method for="WindowSharedStorage">run(|name|, |options|)</dfn> method steps are:
+
+    1. Let |promise| be a new [=promise=].
+    1. If {{Worklet/addModule()}} has not yet been called, reject |promise|.
+    1. Let |outsideSettings| be {{WindowSharedStorage/worklet}}'s [=relevant settings object=].
+    1. Let |agent| be the result of [=obtaining a worklet agent=] given |outsideSettings|.
+    1. Run the following steps in |agent|:
+        1. If {{WindowSharedStorage/worklet}}'s [=module map=] is not [=map/empty=]:
+            1. Let |operationMap| be this {{SharedStorageWorkletGlobalScope}}'s [=SharedStorageWorkletGlobalScope/operation map=].
+            1. If |operationMap| [=map/contains=] |name|:
+                1.  Let |operation| be |operationMap|[|name|].
+                1. If |options| [=map/contains=] |data|:
+                    1. Let |argumentsList| be a new [=list=].
+                    1. [=list/Append=] |data| to |argumentsList|.
+                    1. [=Call=] |operation| with |argumentsList|.
+                1. Otherwise, [=call=] |operation| without any arguments list.
+        1. Resolve |promise| as undefined.
+    1. Return |promise|.
+  </div>
+
+  <div algorithm>
+    The <dfn method for="WindowSharedStorage">selectURL(|name|, |urls|, |options|)</dfn> method steps are:
+
+    1. Let |configPromise| be a new [=promise=].
+    1. If |urls| is empty or exceeds the maximum allowed length, reject |configPromise|.
+    1. Let |urlList| be an empty {{list}}.
+    1. For each |urlWithMetadata| in |urls|:
+        1. If |urlWithMetadata| has no field "`url`", reject |configPromise|.
+        1. Otherwise, let |urlString| be |urlWithMetadata|["`url`"].
+        1. Let |url| be the result of running a [=URL parser=] on |urlString|.
+        1. If |url| is not a valid {{URL}}, reject |configPromise|.
+        1. Otherwise, [=list/append=] |urlString| to |urlList|.
```

Comment:
Why do we parse the URL if we are going to use the original string?  Should we instead be serializing the parsed url back to a string so we get any normalization that occurred?